sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlGenerator.scala

```diff
@@ -16,164 +16,201 @@
         (0 until v.numElements()).foreach { i =>
-          writeChild(options.arrayElementName, ty, v.get(i, ty))
+          writeChildElement(name, ty, v.get(i, ty))
         }
+      // For other datatypes, we just write normal elements.
+      case _ =>
+        writeChildElement(name, dt, v)
+    }
+  }
 
-      case (MapType(_, vt, _), mv: Map[_, _]) =>
-        val (attributes, elements) = mv.toSeq.partition { case (f, _) =>
-          f.toString.startsWith(options.attributePrefix) && f.toString != options.valueTag
-        }
-        // We need to write attributes first before the value.
-        (attributes ++ elements).foreach { case (k, v) =>
-          writeChild(k.toString, vt, v)
-        }
+  def writeElement(dt: DataType, v: Any, options: XmlOptions): Unit = (dt, v) match {
+    case (_, null) | (NullType, _) => gen.writeCharacters(options.nullValue)
+    case (StringType, v: UTF8String) => gen.writeCharacters(v.toString)
+    case (StringType, v: String) => gen.writeCharacters(v)
+    case (TimestampType, v: Timestamp) =>
+      gen.writeCharacters(options.timestampFormatterInWrite.format(v.toInstant()))
+    case (TimestampType, v: Long) =>
+      gen.writeCharacters(options.timestampFormatterInWrite.format(v))
+    case (DateType, v: Int) =>
+      gen.writeCharacters(options.dateFormatterInWrite.format(v))
+    case (IntegerType, v: Int) => gen.writeCharacters(v.toString)
+    case (ShortType, v: Short) => gen.writeCharacters(v.toString)
+    case (FloatType, v: Float) => gen.writeCharacters(v.toString)
+    case (DoubleType, v: Double) => gen.writeCharacters(v.toString)
+    case (LongType, v: Long) => gen.writeCharacters(v.toString)
+    case (DecimalType(), v: java.math.BigDecimal) => gen.writeCharacters(v.toString)
+    case (DecimalType(), v: Decimal) => gen.writeCharacters(v.toString)
+    case (ByteType, v: Byte) => gen.writeCharacters(v.toString)
+    case (BooleanType, v: Boolean) => gen.writeCharacters(v.toString)
+
+    // For the case roundtrip in reading and writing XML files, [[ArrayType]] cannot have
+    // [[ArrayType]] as element type. It always wraps the element with [[StructType]]. So,
+    // this case only can happen when we convert a normal [[DataFrame]] to XML file.
+    // When [[ArrayType]] has [[ArrayType]] as elements, it is confusing what is element name
+    // for XML file.
+    case (ArrayType(ty, _), v: ArrayData) =>
+      (0 until v.numElements()).foreach { i =>
+        writeChild(options.arrayElementName, ty, v.get(i, ty))
+      }
 
-      case (mt: MapType, mv: MapData) => writeMapData(mt, mv)
+    case (MapType(_, vt, _), mv: Map[_, _]) =>
+      val (attributes, elements) = mv.toSeq.partition { case (f, _) =>
+        f.toString.startsWith(options.attributePrefix) && f.toString != options.valueTag
+      }
+      // We need to write attributes first before the value.
+      (attributes ++ elements).foreach { case (k, v) =>
+        writeChild(k.toString, vt, v)
```

Comment:
This may out of scope for this project, or you may be constrained by existing files of this format.

Still, I wanted to mention that map keys are part of the data, and as a result you can end up writing invalid XML files. For example:
```
drop table if exists xml_map_bad;

create table xml_map_bad using xml options (rowTag "row") as
select (map('1', 1, '2', 2, '3', 2)) as map1;
```
The resulting file contains this XML:
```XML
<ROWS>
    <row>
        <map1>
            <1>1</1>
            <2>2</2>
            <3>2</3>
        </map1>
    </row>
</ROWS>
```
Tag names should not start with a digit (as I understand the rules, assuming I am not mistaken).

Querying this table demonstrates the issue.
```
spark-sql (default)> select * from xml_map_bad;
NULL
Time taken: 0.094 seconds, Fetched 1 row(s)
spark-sql (default)> 

```
Similar issues can happen with column names or structure field names. For example:
```
drop table if exists xml_funny_colname;

create table xml_funny_colname
(`funny name 1` int, `funny name 2` int)
using xml options (rowTag "row");

insert into xml_funny_colname
select 1, 2;
```
The tag names contain blanks
```XML
<ROWS>
    <row>
        <funny name 1>1</funny name 1>
        <funny name 2>2</funny name 2>
    </row>
</ROWS>
```
This appears to be an issue.
```
spark-sql (default)> select * from xml_funny_colname;
NULL	NULL
Time taken: 2.24 seconds, Fetched 1 row(s)
spark-sql (default)> 
```
At least, _something_ has gone wrong. I tried to parse the resulting XML:
```
import java.io.File
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.DocumentBuilder
import org.w3c.dom.Document

val factory = DocumentBuilderFactory.newInstance();
// factory.setValidating(true)
factory.setIgnoringElementContentWhitespace(true)

val builder = factory.newDocumentBuilder()

val dir = new File("./spark-warehouse/xml_funny_colname")
val file = dir.listFiles.filter(_.getName.startsWith("part-")).head 

val doc = builder.parse(file)
```
Running the above code results in the error message:
```
org.xml.sax.SAXParseException: Attribute name "name" associated with an element type "funny" must be followed by the ' = ' character.
```

Reply:
cc @sandip-db

Reply:
@bersprockets Thanks for sharing the issue. This came up during our internal testing as well. I can't think of a general way to address all invalid XML element names. But I will add an option that would let users enable validation check for XML element names.